src/test/app/MPToken_test.cpp

```diff
@@ -0,0 +1,1976 @@
+            Account const alice{"alice"};
+            Account const bob{"bob"};
+            Account const carol{"carol"};
+            env.fund(XRP(1000), carol);
+            env.close();
+            MPTTester mptAlice(env, alice, {.holders = {bob}});
+
+            auto const fakeMpt = ripple::test::jtx::MPT(
+                alice.name(), makeMptID(env.seq(alice), alice));
+
+            // issuer tries to clawback MPT where issuance doesn't exist
+            env(claw(alice, fakeMpt(5), bob), ter(tecOBJECT_NOT_FOUND));
+            env.close();
+
+            // alice creates issuance
+            mptAlice.create(
+                {.ownerCount = 1, .holderCount = 0, .flags = tfMPTCanClawback});
+
+            // alice tries to clawback from someone who doesn't have MPToken
+            mptAlice.claw(alice, bob, 1, tecOBJECT_NOT_FOUND);
+
+            // bob creates a MPToken
+            mptAlice.authorize({.account = bob});
+
+            // clawback fails because bob currently has a balance of zero
+            mptAlice.claw(alice, bob, 1, tecINSUFFICIENT_FUNDS);
+
+            // alice pays bob 100 tokens
+            mptAlice.pay(alice, bob, 100);
+
+            // carol fails tries to clawback from bob because he is not the
+            // issuer
+            mptAlice.claw(carol, bob, 1, tecNO_PERMISSION);
+        }
+
+        // clawback more than max amount
+        // fails in the json parser before
+        // transactor is called
+        {
+            Env env(*this, features);
+            Account const alice{"alice"};
+            Account const bob{"bob"};
+
+            env.fund(XRP(1000), alice, bob);
+            env.close();
+
+            auto const mpt = ripple::test::jtx::MPT(
+                alice.name(), makeMptID(env.seq(alice), alice));
+
+            Json::Value jv = claw(alice, mpt(1), bob);
+            jv[jss::Amount][jss::value] = to_string(maxMPTokenAmount + 1);
+            Json::Value jv1;
+            jv1[jss::secret] = alice.name();
+            jv1[jss::tx_json] = jv;
+            auto const jrr = env.rpc("json", "submit", to_string(jv1));
+            BEAST_EXPECT(jrr[jss::result][jss::error] == "invalidParams");
+        }
+    }
+
+    void
+    testClawback(FeatureBitset features)
+    {
+        testcase("MPT Clawback");
+        using namespace test::jtx;
+
+        {
+            Env env(*this, features);
+            Account const alice{"alice"};
+            Account const bob{"bob"};
+
+            MPTTester mptAlice(env, alice, {.holders = {bob}});
+
+            // alice creates issuance
+            mptAlice.create(
+                {.ownerCount = 1, .holderCount = 0, .flags = tfMPTCanClawback});
+
+            // bob creates a MPToken
+            mptAlice.authorize({.account = bob});
+
+            // alice pays bob 100 tokens
+            mptAlice.pay(alice, bob, 100);
+
+            mptAlice.claw(alice, bob, 1);
+
+            mptAlice.claw(alice, bob, 1000);
```

Comment:
Could you verify the balances (for both the issuance and bob) after each of the clawbacks?

Reply:
The check is bundled into `MPTTester::claw()`

     env_.require(
        mptbalance(*this, issuer, issuerAmt - std::min(holderAmt, amount)));
     env_.require(
        mptbalance(*this, holder, holderAmt - std::min(holderAmt, amount)));

Reply:
Ah. Ok. I forgot about that. That's the mixed blessing of these very-smart worker / helper functions.

Would it be worth renaming them? I don't know what a good name would be, though.

Reply:
I think all of those helpers do the necessary checks. Perhaps except for `destroy()`. So maybe don't need to rename, otherwise have to rename all?

Reply:
I don't think we need to rename for the reason Greg stated.